moshi/src/main/java/com/squareup/moshi/BufferedSourceJsonReader.java

```diff
@@ -562,12 +566,60 @@ private boolean isLiteral(int c) throws IOException {
       return -1;
     }
 
-    int result = source.select(options.doubleQuoteSuffix);
-    if (result != -1) {
-      peeked = PEEKED_NONE;
-      pathNames[stackSize - 1] = options.strings[result];
+    for (int i = 0, size = options.doubleQuoteSuffix.size(); i < size; i++) {
+      ByteString bytes = options.doubleQuoteSuffix.get(i);
+
+      if (buffer.size() >= bytes.size() && skipIfNextBufferBytesEqualTo(bytes)) {
+        peeked = PEEKED_NONE;
+        pathNames[stackSize - 1] = options.strings[i];
+
+        return i;
+      }
     }
-    return result;
+
+    return -1;
+  }
+
+  /**
+   * Returns {@code true} if the underlying buffer contains the same bytes as in {@code option} and
+   * skips them in the buffer, otherwise {@code false} (no bytes will be skipped).
+   */
+  private boolean skipIfNextBufferBytesEqualTo(ByteString option) throws IOException {
+    int skipBytes = 0; // We can't rely on the for loop, counting separately
+    for (int i = 0, size = option.size(); i < size; i++) {
+      byte b = buffer.getByte(skipBytes);
+      if (b != option.getByte(i)) {
+        // The character may be escaped
+        if (b == '\\') {
+          // The buffer should have enough bytes for a char.
+          if (buffer.size() >= i + ESCAPE_BYTES_SIZE && buffer.getByte(i + 1) == 'u') {
+            byte[] hex = new byte[4];
+            int offset = 2;
+            while (offset != ESCAPE_BYTES_SIZE) {
+              hex[offset - 2] = buffer.getByte(i + offset);
+              offset++;
+            }
+
+            int value;
+            try {
+              value = Integer.parseInt(ByteString.of(hex).utf8(), 16);
```

Comment:
I think this might be a decent strategy.

1. First attempt `source.select()`. If that works, we’re good.
2. Next call `nextString()`.
3. Check if the string matches any of the options. If it does, we’re done.
4. Push the consumed string back by setting `peeked` and `peekedString`. See `promoteNameToValue()` for an example.

The tricky part is the weird combinations of calls that should always leave `JsonReader` in a consistent state. It’s difficult ’cause we coerce types.

```
    JsonReader reader = newReader("[123]");
    reader.beginArray();
    assertEquals(-1, reader.selectString(JsonReader.Options.of("124")));
    assertEquals(123, reader.nextLong());
```